curvefs/src/metaserver/inode_manager.cpp

```diff
@@ -308,23 +315,44 @@ MetaStatusCode InodeManager::UpdateInode(const UpdateInodeRequest &request) {
 MetaStatusCode InodeManager::GetOrModifyS3ChunkInfo(
     uint32_t fsId, uint64_t inodeId,
     const S3ChunkInfoMap& map2add,
-    std::shared_ptr<Iterator>* iterator,
+    const S3ChunkInfoMap& map2del,
     bool returnS3ChunkInfoMap,
-    bool compaction) {
+    std::shared_ptr<Iterator>* iterator) {
     VLOG(1) << "GetOrModifyS3ChunkInfo, fsId: " << fsId
             << ", inodeId: " << inodeId;
 
     NameLockGuard lg(inodeLock_, GetInodeLockName(fsId, inodeId));
 
-    if (!map2add.empty()) {
-        for (const auto& item : map2add) {
-            uint64_t chunkIndex = item.first;
-            auto list2add = item.second;
-            MetaStatusCode rc = inodeStorage_->AppendS3ChunkInfoList(
-                fsId, inodeId, chunkIndex, list2add, compaction);
-            if (rc != MetaStatusCode::OK) {
-                return rc;
-            }
+    std::unordered_map<uint64_t, bool> deleted;
+    for (const auto& item : map2add) {
+        uint64_t chunkIndex = item.first;
+        auto list2add = item.second;
+        S3ChunkInfoList list2del;
+        auto iter = map2del.find(chunkIndex);
+        if (iter != map2del.end()) {
+            list2del = iter->second;
+        }
+
+        MetaStatusCode rc = inodeStorage_->ModifyInodeS3ChunkInfoList(
+            fsId, inodeId, chunkIndex, list2add, list2del);
+        if (rc != MetaStatusCode::OK) {
+            return rc;
+        }
+        deleted[chunkIndex] = true;
+    }
+
+    for (const auto& item : map2del) {
+        uint64_t chunkIndex = item.first;
+        if (deleted[chunkIndex]) {
+            continue;
+        }
+
+        auto list2del = item.second;
+        S3ChunkInfoList dummy;
```

Comment:
it's wired that you add a empty list, even now it's correct because `ModifyInodeS3ChunkInfoList` will just return ok if list is empty.

Reply:
> it's wired that you add a empty list, even now it's correct because `ModifyInodeS3ChunkInfoList` will just return ok if list is empty.

yep, i will use pointer to improve it.